### PR TITLE
Test controllers topic+sample updates

### LINKED ISSUE #8147: ActionResult<T> controller method example
How do I test a controller method returning ActionResult&lt;T&gt; or Task&lt;ActionResult&lt;T&gt;&gt; in 2.1?

Thanks

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b5cce372-24e2-3489-cb19-c85d934e5d3b
* Version Independent ID: 9d12b19f-a94e-4df4-357b-3f3b93696eb6
* Content: [Test controller logic in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/testing?view=aspnetcore-2.1)
* Content Source: [aspnetcore/mvc/controllers/testing.md](https://github.com/aspnet/Docs/blob/master/aspnetcore/mvc/controllers/testing.md)
* Product: **aspnet-core**
* GitHub Login: @ardalis
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Addresses #5495 
Fixes #8147 

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/controllers/testing?view=aspnetcore-1.0&branch=pr-en-us-8232)

cc @mberginbh

### CODE REVIEW

**Comment by guardrex** [comment]
@javiercn Are you available to take quick look at this comment? There's only one significant change to this topic, and I can paste it right into this comment. The change is implementing a return of an `ActionResult<T>` and two tests against it. Everything else is pretty much unchanged.

Here are the relevant bits ...

> In ASP.NET Core 2.1 or later, `ActionResult<T>` type (`ActionResult<TValue>`) enables you to return a type deriving from `ActionResult` or return a specific type. The sample app includes a method that returns a `List<IdeaDTO>` for a given session `id`:

NOTE: In that topic text :point_up:, `ActionResult<T>` is linked to topic content, while `ActionResult<TValue>` is linked to the API.

```csharp
[HttpGet("forsessionactionresult/{sessionId}")]
public async Task<ActionResult<List<IdeaDTO>>> ForSessionActionResult(int sessionId)
{
    var session = await _sessionRepository.GetByIdAsync(sessionId);

    if (session == null)
    {
        return NotFound(sessionId);
    }

    var result = session.Ideas.Select(idea => new IdeaDTO()
        {
            Id = idea.Id,
            Name = idea.Name,
            Description = idea.Description,
            DateCreated = idea.DateCreated
        }).ToList();

    return result;
}
```

> Two tests of the `ForSessionActionResult` controller are included in the `ApiIdeasControllerTests`. The first test confirms that the method returns an `ActionResult`, session, and idea for a valid session `id`:
> 
> * The `ActionResult` type is `ActionResult<List<IdeaDTO>>`.
> * The value of the `ActionResult` is a `List<IdeaDTO>`.
> * The first item in the list is a valid `IdeaDTO` matching the idea added to the mock session.

```csharp
[Fact]
public async Task ForSessionActionResult_ReturnsIdeasForSession()
{
    // Arrange
    int testSessionId = 123;
    var mockRepo = new Mock<IBrainstormSessionRepository>();
    mockRepo.Setup(repo => repo.GetByIdAsync(testSessionId))
        .Returns(Task.FromResult(GetTestSession()));
    var controller = new IdeasController(mockRepo.Object);

    // Act
    var result = await controller.ForSessionActionResult(testSessionId);

    // Assert
    var actionResult = Assert.IsType<ActionResult<List<IdeaDTO>>>(result);
    var returnValue = Assert.IsType<List<IdeaDTO>>(actionResult.Value);
    var idea = returnValue.FirstOrDefault();
    Assert.Equal("One", idea.Name);
}
```

> The second test confirms that the controller returns an `ActionResult` but not a nonexistent session for a bad session `id`:
> 
> * The `ActionResult` type is `ActionResult<List<IdeaDTO>>`.
> * The `ActionResult.Result` is a `NotFoundObjectResult`.

```csharp
[Fact]
public async Task ForSessionActionResult_ReturnsNotFoundObjectResultForNonexistentSession()
{
    // Arrange
    var mockRepo = new Mock<IBrainstormSessionRepository>();
    var controller = new IdeasController(mockRepo.Object);
    var nonExistentSessionId = 999;

    // Act
    var result = await controller.ForSessionActionResult(nonExistentSessionId);

    // Assert
    var actionResult = Assert.IsType<ActionResult<List<IdeaDTO>>>(result);
    Assert.IsType<NotFoundObjectResult>(actionResult.Result);
}
```

That's it. If that looks ok, we should be good to take this PR into a UE review.

... and if you're thinking that the method/test naming is a little funky ("ActionResult" in the names), that's probably ok. This is just a demo sample app, and those names are close to the non-`ActionResult<T>` bits.

**Comment by guardrex** [comment]
@scottaddie 

I gave this topic a haircut to focus it more on testing and a bit less on controller architecture advice. If you see some things we should retain, I can either reconstitute them back into the topic or move them over to the controllers overview topic.

**Comment by guardrex** [comment]
One thing here occurs to me to check ...

It's fine that in the "not found" case that it looks at the `ActionResult<T>.Result` to confirm it's a `NotFoundObjectResult`. It's fine that the returned `List<IdeaDTO>` case checks the `ActionResult<T>.Value`. What I didn't check was the `ActionResult<T>.Result` in the `List<IdeaDTO>` case. I think that should happen right before the `ActionResult<T>.Value` is checked.

I'm going to look at that right now and see what's what and report back.

**Comment by guardrex** [comment]
Ok, well it comes back `null`, and I don't think we'd test for that. Apparently, when there are entities to return, the `ActionResult<T>.Result` isn't set. When there isn't anything to return, it gets a `NotFoundObjectResult`. I think this is ok the way I have it (*admittedly guessing a bit*).

**Comment by guardrex** [comment]
Added a method for POST (`CreateActionResult`) and three tests that mirror the `IActionResult` approach.

I haven't updated the text of the topic yet. Let's discuss these last additions and make sure they're good first.

Also, still not clear on testing for 200/201 when the method is successful. Looks like `ActionResult<T>.Result` isn't populated when the method succeeds. Therefore, I'm not clear yet on testing the status code, too. It's not really very necessary given that the methods return objects that can be checked, but it would be good to know if it's possible to test for status codes, too.

**Comment by guardrex** [comment]
@pascalpfeil suggested some improvements, which I placed into the last commit ...

* `CreatedAtAction`
* Test checks *201 Created* (`CreatedAtActionResult`) and value.

Still don't know about checking *200 Ok* for an `ActionResult<T>` success tho ...

```csharp
// Assert
var actionResult = Assert.IsType<ActionResult<List<IdeaDTO>>>(result);

Assert.IsType<OkObjectResult>(actionResult.Result);

var returnValue = Assert.IsType<List<IdeaDTO>>(actionResult.Value);
var idea = returnValue.FirstOrDefault();
Assert.Equal("One", idea.Name);
```

`ActonResult<T>.Result` is `null`, so that doesn't work.

**Comment by guardrex** [comment]
... and I haven't updated the text for the latest code examples. I'll do that on the next commit.

**Comment by guardrex** [comment]
@scottaddie Ok ... there's some real *meat on the bone* 🍖 here. I've updated based on your feedback, and I've updated/added the explanations of the controller methods and the tests of those methods.

Only one thing is still unclear to me ...

Idk how to get an `OkObjectResult` when the GET method successfully returns the list of requested objects ...

```csharp
// Assert
var actionResult = Assert.IsType<ActionResult<List<IdeaDTO>>>(result);

Assert.IsType<OkObjectResult>(actionResult.Result);

var returnValue = Assert.IsType<List<IdeaDTO>>(actionResult.Value);
var idea = returnValue.FirstOrDefault();
Assert.Equal("One", idea.Name);
```

`ActonResult<T>.Result` is `null`, so that doesn't work.

Now, that doesn't really stop us from going ahead and merging the PR. Still tho ... if a 200 comes back with the list of objects, then imo there should be a way to test that along with the objects returned. We can assume that the objects signify a *200 Ok*, but it would be nice to test the status code, too, imo.

File: aspnetcore/mvc/controllers/testing/sample/TestingControllersSample/tests/TestingControllersSample.Tests/UnitTests/ApiIdeasControllerTests.cs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Net;
```
Comment (scottaddie) [inline aspnetcore/mvc/controllers/testing/sample/TestingControllersSample/tests/TestingControllersSample.Tests/UnitTests/ApiIdeasControllerTests.cs]: Remove this unused reference.

**Comment by scottaddie** [comment]
@guardrex I'm researching your 200 OK question right now. Please also set the 2.1 compat flag in `Startup.ConfigureServices`.

**Comment by scottaddie** [comment]
@guardrex I chatted with @rynowak about it. The action under test would have to return `Ok(object)` to access the object result. In this case, the action returns a generic List, so that's what the unit test sees. You should remove `Assert.IsType<OkObjectResult>(actionResult.Result);` from the test. Let's have Ryan do an engineering review of this PR when you're ready.

**Comment by guardrex** [comment]
> You should remove Assert.IsType<OkObjectResult>(actionResult.Result); from the test.

Yes, that was never on a commit. That was just shot in here for discussion.

I'll make your updates and then call for the review. Thanks for helping with this. :beers:

**Comment by Rick-Anderson** [comment]
@javiercn please review or recommend someone to review or provide a review ETA.

File: aspnetcore/mvc/controllers/testing/sample/TestingControllersSample/tests/TestingControllersSample.Tests/UnitTests/ApiIdeasControllerTests.cs [hunk 2]
```diff
@@ -117,6 +127,130 @@ public async Task ForSession_ReturnsIdeasForSession()
...
+
+        #region snippet_ForSessionActionResult_ReturnsIdeasForSession
+        [Fact]
+        public async Task ForSessionActionResult_ReturnsIdeasForSession()
+        {
+            // Arrange
+            int testSessionId = 123;
+            var mockRepo = new Mock<IBrainstormSessionRepository>();
+            mockRepo.Setup(repo => repo.GetByIdAsync(testSessionId))
+                .Returns(Task.FromResult(GetTestSession()));
```
Comment (scottaddie) [inline aspnetcore/mvc/controllers/testing/sample/TestingControllersSample/tests/TestingControllersSample.Tests/UnitTests/ApiIdeasControllerTests.cs]: Use `ReturnsAsync` instead. It's a bit cleaner. Consider that change all throughout the sample. In this example, the code would become `.ReturnsAsync(GetTestSession());`.

**Comment by guardrex** [comment]
@scottaddie Got it. Anything similarly better for ...

```csharp
.Returns(Task.CompletedTask)
```

**Comment by scottaddie** [comment]
@guardrex I'm not aware of a better way to handle that.

**Comment by guardrex** [comment]
We still need engineering to take a look at this?

I'm anxious to get all of my PRs merged by next Tuesday. I'm a big 'ole :skull: after that. :smile: lol

**Comment by scottaddie** [comment]
@guardrex If you don't hear anything from engineering the end of today, go ahead and merge.

**Comment by guardrex** [comment]
@scottaddie Ok, thanks. Will do.

**Pull request merged**